src/include/mpir_objects.h

```diff
@@ -303,23 +303,43 @@ typedef MPL_atomic_int_t Handle_ref_count;
         HANDLE_LOG_REFCOUNT_CHANGE(objptr_, new_ref_, "incr");          \
         HANDLE_CHECK_REFCOUNT(objptr_,new_ref_,"incr");                 \
     } while (0)
-#define MPIR_Object_release_ref_always(objptr_,inuse_ptr)               \
-    do {                                                                \
-        int new_ref_ = MPL_atomic_fetch_sub_int(&((objptr_)->ref_count), 1) - 1; \
-        *(inuse_ptr) = new_ref_;                                        \
-        HANDLE_LOG_REFCOUNT_CHANGE(objptr_, new_ref_, "decr");          \
-        HANDLE_CHECK_REFCOUNT(objptr_,new_ref_,"decr");                 \
+#define MPIR_Object_release_ref_always(objptr_,inuse_ptr)                      \
+    do {                                                                       \
+        /* If it is 1, we will just free it without a heavy atomic operation.  \
+         * Note that any concurrent add_ref() to a handle whose count is 1 is  \
+         * illegal and we do not consider.                                     \
+         * (Rationale: if fetch_add() and fetch _sub() are called in parallel, \
+         *  fetch_sub() can happen to be executed earlier than fetch_add(),    \
+         *  which frees this object.  The caller of fetch_add() thinks this    \
+         * object is alive, causing invalid access. */                         \
```

Comment:
Personally I don't think we need "Rationale" comment in the code. They can be in the commit messages. Rationale is for understanding related to background knowledge or reasoning. I think the comment of the case being "illegal" is sufficient for reading the code.